File: src/sites/HentaiFoundry.ts
import BaseSite from './BaseSite';
import ProcessResponse from './ProcessResponse';
import { CookieJar } from 'tough-cookie';
import cheerio from 'cheerio';
import got from 'got';
import { Message, MessageEmbed } from 'discord.js';
import { DateTime } from 'luxon';
import CacheManager from '../CacheManager';
import { processDescription } from '../Helpers';
import { HENTAI_FOUNDRY_ICON_URL } from '../Constants';

class HentaiFoundry extends BaseSite {
    identifier = 'Hentai Foundry';

    pattern =
        /https:?\/\/(www\.)?hentai-foundry\.com\/pictures\/user\/(?<user>.*)\/(?<id>\d+)\/(?<slug>\S+)\/?/gim;

    color = 0xff67a2;

    baseUrl = 'https://www.hentai-foundry.com';

    jar: CookieJar;

    constructor() {
        super();

        this.jar = new CookieJar();
    }

    async process(
        match: RegExpMatchArray,
        /* eslint-disable @typescript-eslint/no-unused-vars */
        source: Message | null
    ): Promise<ProcessResponse | false> {
        const message: ProcessResponse = {
            embeds: [],
            files: [],
        };

        if (!match.groups?.id || !match.groups?.slug) {
            return Promise.resolve(false);
        }

        await got.get(`${this.baseUrl}/?enterAgree=1`, { cookieJar: this.jar });

        const url = match[0];

        const response = await this.getPage(
            url,
            match.groups.id,
            match.groups.slug
        );

        if (!response) {
            return Promise.resolve(false);
        }

        const $ = cheerio.load(response);

        const title = $('.imageTitle');
        const description = $('.picDescript');
        const image = $('#picBox .boxbody img');

        // We basically rely on the fact the posted at field is going to use this time element.
        const postedAt = $('#pictureGeneralInfoBox time');

        // HF has no specific selectors for these elements, so we do this jank method
        // That involves finding the views label and then getting the next text node's content and trimming it
        const views = $(
            '#pictureGeneralInfoBox .boxbody .column span:contains("Views")'
        )?.[0]?.next?.data?.trim();

        const votes = $(
            '#pictureGeneralInfoBox .boxbody .column span:contains("Vote Score")'
        )?.[0]?.next?.data?.trim();

        const authorLink = $('#descriptionBox .boxbody a');
        const authorImage = $('#descriptionBox .boxbody a img');

        const embed = new MessageEmbed({
            title: title.text(),
            url: url,
            timestamp: DateTime.fromISO(postedAt?.attr('datetime') ?? '')
                .toUTC()
                .toMillis(),
            description: processDescription(description.text()),
            color: this.color,
            image: {
                url: `https:${image.attr('src')}`,
            },
            author: {
                name: authorImage.attr('title'),
                url: `${this.baseUrl}${authorLink.attr('href')}`,
                iconURL: `https:${authorImage.attr('src')}`,
            },
            fields: [
                {
                    name: 'Views',
                    value: views ?? '0',
                    inline: true,
                },
                {
                    name: 'Votes',
                    value: votes ?? '0',
                    inline: true,
                },
            ],
            footer: {
                iconURL: HENTAI_FOUNDRY_ICON_URL,
                text: 'HentaiFoundry',
            },
        });

        message.embeds.push(embed);

        return Promise.resolve(message);
    }

    async getPage(
        url: string,
        id: string,
        slug: string
    ): Promise<string | null> {
        const cacheKey = `hentaifoundry.picture_${id}_${slug}`;
        const cacheManager = await CacheManager.getInstance();

        const cachedValue = await cacheManager.remember(cacheKey, async () => {
            const response = await got.get(url, { cookieJar: this.jar });
            return Promise.resolve(response.body);
        });

        if (!cachedValue) {
            return Promise.resolve(null);
        }

        return Promise.resolve(cachedValue);
    }
}

export default HentaiFoundry;

File: SaucyBot/Site/HentaiFoundry.cs
using System.Text.RegularExpressions;
using Discord;
using Discord.WebSocket;
using SaucyBot.Common;
using SaucyBot.Library;
using SaucyBot.Library.Sites.HentaiFoundry;
using SaucyBot.Site.Response;

namespace SaucyBot.Site;

public class HentaiFoundry : BaseSite
{
    public override string Identifier => "HentaiFoundry";

    protected override string Pattern =>
        @"https:?\/\/(www\.)?hentai-foundry\.com\/pictures\/user\/(?<user>.*)\/(?<id>\d+)\/(?<slug>\S+)";

    protected override Color Color => new(0xFF67A2);

    private readonly ILogger<HentaiFoundry> _logger;
    private readonly HentaiFoundryClient _client;

    public HentaiFoundry(ILogger<HentaiFoundry> logger, HentaiFoundryClient client)
    {
        _logger = logger;
        _client = client;
    }

    public override async Task<ProcessResponse?> Process(Match match, SocketUserMessage? message = null)
    {
        var response = new ProcessResponse();

        if (!await _client.Agree())
        {
            _logger.LogError("HentaiFoundry over 18 agreement failed, cookie was not present");
            return null;
        }

        var page = await _client.GetPage(match.Value);

        if (page is null)
        {
            return response;
        }

        var embed = new EmbedBuilder
        {
            Title = page.Title(),
            Description = await Helper.ProcessDescription(page.Description() ?? ""),
            Url = match.Value,
            Color = this.Color,
            ImageUrl = page.ImageUrl(),
            Timestamp = page.PostedAt(),
            Author = new EmbedAuthorBuilder
            {
                Name = page.AuthorName(),
                Url = page.AuthorUrl(),
                IconUrl = page.AuthorAvatarUrl(),
            },
            Fields = new List<EmbedFieldBuilder>
            {
                new()
                {
                    Name = "Views",
                    Value = page.Views(),
                    IsInline = true,
                },
                new()
                {
                    Name = "Votes",
                    Value = page.Votes(),
                    IsInline = true,
                }
            },
            Footer = new EmbedFooterBuilder { IconUrl = Constants.HentaiFoundryIconUrl, Text = "HentaiFoundry" },
        };
        
        response.Embeds.Add(embed.Build());

        return response;
    }
}
